mapper/javabean/src/main/java/org/hibernate/search/mapper/javabean/loading/IndexLoader.java

```diff
@@ -0,0 +1,21 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.mapper.javabean.loading;
+
+import java.util.List;
+
+public interface IndexLoader<E> {
+
+	/**
+	 * Loads the entities corresponding to the given identifiers, blocking the current thread while doing so.
+	 *
+	 * @return A list of entities, in the same order the identifiers were given.
+	 * {@code null} is inserted when an object is not found or has the wrong concrete type.
+	 */
+	List<E> load();
```

Comment:
I don't understand why this would load all entities; this seems awfully inefficient. In some scenarios where there are millions of entities persisted on disk, this would even trigger an out of memory error.

The worst part is, in the implementation you apparently convert the entities to identifiers, and then load entities by id! Which is something that `EntityLoader` is perfectly able to do.

The strategy in the ORM mass indexer was to load identifiers one after another by batches, and to load entities by batches, so that we never have more than a batch's worth of IDs/entities in memory at any given time. That worked quite well even for huge data sets (dozens of millions of entities).

I suggest you take a similar approach here. I would define the API like this:

* Remove this `IndexLoader` interface.
* Add a `EntityIdentifierScroll` interface, extending `AutoCloseable`, with two methods: `void close():` and `List<?> next()`.
* Add a `EntityIdentifierLoader` interface, with one methods: `EntityIdentifierScroll allIdentifiers():`.
* In `LoadingOptions`, replace `registerIndexLoader(Class<T> type, IndexLoader<T> loader` with `registerIdentifierLoader(Class<?>, EntityIdentifierLoader)`.
* Rely on the user-provided `EntityLoader` and `EntityIdentifierLoader` to implement `org.hibernate.search.mapper.javabean.loading.impl.JavaBeanMassIndexingTypeLoadingStrategy.JavaBeanMassIndexingTypeLoader#loadIndex`.

Reply:
Currently, I was building my third module that uses mapperpojo as composites to build my own mapper. The point is that everyone, having at their disposal mapperpojo, can build their own module. But also, mapper-javabean was such a basic mapping for non-JPA objects.

I was thinking of making changes to mapperjavabean in the next step. Building a load and indexing strategy "SearchMapping.createSessionWithOptions ()" is not good. In addition, it is terribly complicated to use elements from SPI. Better to add indexing strategies and entity loading strategies while building the mapper in "SearchMappingBuilder". 

Zbudowałem w SPI specjalny DSL do ładowania wsadowego. Użyty jest w mapper-orm i mapper-javabean. Jest tak samo jak w mapper-orm, jako strategia ładowania, tak samo strategia masowej indeksacji.    W mapper-orma wszystkie strategie ładuje w SPI. W mapping-javabean niedawno ładowanie encji weszło od strony API.

But DSL for index loading is ready. I used it in hibernate-orm replacing the previous solution but keeping all functionality, remembering about batch loading in "org.hibernate.search.mapper.orm.loading.impl.HibernateOrmJpaMassIndexingTypeLoadingStrategy". There is also a similar module in mapper-javabean using this inefficient bridge in "org.hibernate.search.mapper.javabean.loading.impl.JavaBeanMassIndexingTypeLoadingStrategy. The batch loading element is" MassIndexingTypeLoader ", it has the ability to load with packets. 

```java
	private static class JavaBeanMassIndexingTypeLoader<E> implements MassIndexingTypeLoader<E, JavaBeanMassIndexerContext> {
		private final JavaBeanMassIndexingTypeLoadingStrategy strategy;
		private final IndexLoader<E> indexLoader;

		public JavaBeanMassIndexingTypeLoader(JavaBeanMassIndexingTypeLoadingStrategy strategy, IndexLoader<E> indexLoader) {
			this.strategy = strategy;
			this.indexLoader = indexLoader;
		}

		@Override
		public void loadIndex(MassIndexingTypeLoadingContext<E, JavaBeanMassIndexerContext> loading) throws InterruptedException {
			MassIndexingSessionContext sessionContext = (MassIndexingSessionContext) loading
					.contextData().get( MassIndexingSessionContext.class );
			List<E> entiteies = indexLoader.load();
			Map<Object, E> map;
			map = entiteies.stream()
					.collect( Collectors.toMap( entity -> strategy.identifier( sessionContext, loading.commonSuperType(), entity ),
							Function.identity() ) );

			MassIndexingTypeBatchingStep batching = loading.batching( (ctx, ids) -> {
				return ids.stream().map( map::get ).collect( Collectors.toList() );
			} );

			loading.totalCount( map.size() );
			int batchSize = loading.indexer().batchSize();
			List<Object> destination = new ArrayList<>( batchSize );
			Set<Object> identifiers = map.keySet();
			for ( Object identifier : identifiers ) {
				destination.add( identifier );
				if ( !loading.isActive() ) {
					throw new InterruptedException();
				}

				if ( destination.size() == batchSize ) {
					batching.load( destination );
					destination.clear();
				}

			}

			batching.load( destination );
		}
	}
```
But still, maybe it is better to implement this package eventually when building in SearchMappingBuilder.

Reply:
Maybe "PojoRawTypeIdentifier" can be moved from SPI to API? Then mass indexing would also be sent to the API. Only that requires it to be in SPI.

Reply:
> The point is that everyone, having at their disposal mapperpojo, can build their own module

Sure. And then they have to rely on SPI, not API. Because it's a lot of work to guarantee stability of APIs, so we don't want to extend them too much. But you're certainly welcome to rely on SPI.

> In addition, it is terribly complicated to use elements from SPI. Better to add indexing strategies and entity loading strategies while building the mapper in "SearchMappingBuilder".

See :) If it's complicated to use, imagine how complicated it is to expose as a simple API!

> I was thinking of making changes to mapper-javabean in the next step. Building a load and indexing strategy "SearchMapping.createSessionWithOptions ()" is not good

I agree, that will have to change. But I don't see how this answers my comment.

**Could you explain, in one simple sentence, why my suggestion would not work?**
As a reminder, this was the suggestion:

>  I suggest you take a similar approach here. I would define the API like this:
> 
> * Remove this `IndexLoader` interface.
> * Add a `EntityIdentifierScroll` interface, extending `AutoCloseable`, with two methods: `void close():` and `List<?> next()`.
> * Add a `EntityIdentifierLoader` interface, with one methods: `EntityIdentifierScroll allIdentifiers():`.
> * In `LoadingOptions`, replace `registerIndexLoader(Class<T> type, IndexLoader<T> loader` with `registerIdentifierLoader(Class<?>, EntityIdentifierLoader)`.
> * Rely on the user-provided `EntityLoader` and `EntityIdentifierLoader` to implement `org.hibernate.search.mapper.javabean.loading.impl.JavaBeanMassIndexingTypeLoadingStrategy.JavaBeanMassIndexingTypeLoader#loadIndex`.

Reply:
I'll think about it differently to get elements into the API. This will simplify the task a lot.

Reply:
Ok.

By the way:

> Maybe "PojoRawTypeIdentifier" can be moved from SPI to API? Then mass indexing would also be sent to the API. Only that requires it to be in SPI.

You don't need `PojoRawTypeIdentifer` for entity loading. You are dealing with entities exclusively, so you can simply use the entity name to refer to an entity.

Reply:
The concept was simple. To easily define batch indexing with simple DSL. I had a problem with quitting sleep to compile. But I already did. I was able to get it out. So you can easily define batch loading: 

While charging, you have easy access to the batch loading element. It is the programmer who decides how to select identifiers from his source. All information from the indexer is available, how to build loops. 

Example:
```java

options.registerIndexLoader( Book.class, (loading) -> {

	MassIndexingTypeBatchingStep batching = loading.batching( (ctx, ids) -> {
		return ids.stream().map( bookmap::get ).collect( Collectors.toList() );
	} );

	loading.totalCount( bookmap.size() );
	int batchSize = loading.indexer().batchSize();

	List<Object> destination = new ArrayList<>( batchSize );
	for ( Object identifier : bookmap.keySet() ) {
		destination.add( identifier );
		if ( !loading.isActive() ) {
			throw new InterruptedException();
		}

		if ( destination.size() == batchSize ) {
			batching.load( destination );
			destination.clear();
		}

	}

	batching.load( destination );

} ).group( MassIndexingTypeGroup.asIstanceOf() ).documentInterceptor( JavaBeanIndexingInterceptors.withJTATransaction( tm ));
```
In addition, I take the functionality out of grouping types for batch loading. As in mapper-orm with the preset options asNone (), asIstanceOf (). It is also possible to define your own grouping type by implementing the "MassIndexingTypeGroup" interface. 
```java
.group( MassIndexingTypeGroup.asIstanceOf() )
```
Another thing is the ability to define interception when producing IDs or producing index documents. I have defined ready-made intercepts for JTA transactions. But the programmer can also create his own interceptions using simple DSL language, defining a double nested aspect. It just needs to implement "MassIndexingInterceptor". 
```java
.documentInterceptor( JavaBeanIndexingInterceptors.withJTATransaction( tm ))
.identifierInterceptor( JavaBeanIndexingInterceptors.withJTATransaction( tm ))
```

Reply:
Due to the necessity to maintain fully compatible functionality with mapper-orm, such api for mapper-javabean was created, which can be used to define everything that was in mapper-orm. 

Loader in mapper-orm indexing strategy is defined in several elements, consisting of the loader itself, interception and grouping definition. 

```java
	private class HibernateOrmJpaStrategyMassIndexingTypeLoader<E> implements HibernateOrmMassIndexingTypeLoader<E> {
		private final HibernateOrmQueryLoader<? super E, I> typeGroupLoader;

		public HibernateOrmJpaStrategyMassIndexingTypeLoader(HibernateOrmQueryLoader<? super E, I> typeGroupLoader) {
			this.typeGroupLoader = typeGroupLoader;
		}

		@Override
		public void loadIndex(MassIndexingTypeLoadingContext<E, HibernateOrmMassIndexerContext> loading) throws InterruptedException {
			HibernateOrmMassIndexerContext indexer = loading.indexer();
			int fetchSize = indexer.fetchSize();
			int batchSize = indexer.batchSize();
			long objectsLimit = indexer.objectsLimit();

			MassIndexingTypeBatchingStep batching = loading.batching( (ctx, ids) -> {
				CacheMode cacheMode = indexer.cacheMode();
				SessionImplementor session = (SessionImplementor) ctx.contextData().get( SessionImplementor.class );
				Query<? super E> query = typeGroupLoader.createLoadingQuery( session, ID_PARAMETER_NAME )
						.setParameter( ID_PARAMETER_NAME, ids )
						.setCacheMode( cacheMode )
						.setLockMode( LockModeType.NONE )
						.setCacheable( false )
						.setHibernateFlushMode( FlushMode.MANUAL )
						.setFetchSize( ids.size() );
				return query.getResultList();
			} );

			SharedSessionContractImplementor session = (SharedSessionContractImplementor) loading
					.contextData().get( SharedSessionContractImplementor.class );

			long totalCount = typeGroupLoader
					.createCountQuery( session )
					.setCacheable( false ).uniqueResult();

			if ( objectsLimit != 0 && objectsLimit < totalCount ) {
				totalCount = objectsLimit;
			}
			if ( log.isDebugEnabled() ) {
				log.debugf( "going to fetch %d primary keys", (Long) totalCount );
			}
			loading.totalCount( totalCount );

			ArrayList<I> destinationList = new ArrayList<>( batchSize );
			long counter = 0;
			try ( ScrollableResults results = typeGroupLoader.createIdentifiersQuery( session )
					.setCacheable( false )
					.setFetchSize( fetchSize ).scroll( ScrollMode.FORWARD_ONLY ) ) {

				while ( results.next() ) {
					@SuppressWarnings("unchecked")
					I id = (I) results.get( 0 );
					destinationList.add( id );
					if ( destinationList.size() == batchSize ) {
						// Explicitly checking whether the TX is still open; Depending on the driver implementation new ids
						// might be produced otherwise if the driver fetches all rows up-front
						if ( !(loading.isActive() && session.isTransactionInProgress()) ) {
							throw log.transactionNotActiveWhileProducingIdsForBatchIndexing( loading.includedEntityNames() );
						}

						batching.load( destinationList );
						destinationList = new ArrayList<>( batchSize );
					}
					counter++;
					if ( counter == totalCount ) {
						break;
					}
				}
			}
			batching.load( destinationList );
		}
	}
```

I created a group selector in maper-orm, which works identically to the previous one: 
```java
	public MassIndexingTypeGroup group() {
		return (superType1, superType2) -> {
			LoadingIndexedTypeContext<?> commonSuperType1 = typeContextProvider
					.indexedForExactType( (PojoRawTypeIdentifier<E>) superType1 );
			LoadingIndexedTypeContext<?> commonSuperType2 = typeContextProvider
					.indexedForExactType( (PojoRawTypeIdentifier<E>) superType2 );

			EntityPersister entityPersister = commonSuperType1.entityPersister();
			EntityPersister otherEntityPersister = commonSuperType2.entityPersister();
			if ( HibernateOrmUtils.isSuperTypeOf( entityPersister, otherEntityPersister ) ) {
				return TypeJoin.TOP;
			}
			if ( HibernateOrmUtils.isSuperTypeOf( otherEntityPersister, entityPersister ) ) {
				return TypeJoin.NEXT;
			}
			return TypeJoin.NONE;
		};
	}
```

By implementing "MassIndexingInterceptor", I created a transaction control according to the previous function in mapper-orm. For identifiers: 

```java 
ublic class HibernateOrmMassIndexingIdentifierProducerInterceptor implements HibernateOrmMassIndexingInterceptor {
	@Override
	public void procesInterceptor(MassIndexingInterceptorInvocationContext<HibernateOrmMassIndexerContext> ictx) throws Exception {
		HibernateOrmMassIndexerContext indexer = (HibernateOrmMassIndexerContext) ictx.indexer();
		Integer transactionTimeout = indexer.transactionTimeout();
		String tenantId = ictx.tenantId();
		boolean wrapInTransaction = wrapInTransaction();
		if ( wrapInTransaction ) {
			try ( StatelessSession statelessSession = factory.withStatelessOptions()
					.tenantIdentifier( tenantId )
					.openStatelessSession() ) {
				if ( transactionTimeout != null ) {
					transactionManager.setTransactionTimeout( transactionTimeout );
				}
				transactionManager.begin();
				inTransactionWrapper( ictx, statelessSession, tenantId );
				transactionManager.commit();
			}
			// Just let runtime exceptions fall through
			catch (NotSupportedException | SystemException | RollbackException | HeuristicMixedException
					| HeuristicRollbackException e) {
				throw log.massIndexingTransactionHandlingException( e.getMessage(), e );
			}
		}
		else {
			inTransactionWrapper( ictx, null, tenantId );
		}
	}
.........
```
For documents: 
```java
public class HibernateOrmMassIndexingDocumentProducerInterceptor implements HibernateOrmMassIndexingInterceptor {
.....
	@Override
	public void procesInterceptor(MassIndexingInterceptorInvocationContext<HibernateOrmMassIndexerContext> ictx) throws Exception {
		HibernateOrmMassIndexerContext indexer = ictx.indexer();
		CacheMode cacheMode = indexer.cacheMode();
		Integer transactionTimeout = indexer.transactionTimeout();
		String tenantId = ictx.tenantId();

		try ( SessionImplementor upperSession = (SessionImplementor) factory
				.withOptions()
				.tenantIdentifier( tenantId )
				.openSession() ) {
			upperSession.setHibernateFlushMode( FlushMode.MANUAL );
			upperSession.setCacheMode( cacheMode );
			upperSession.setDefaultReadOnly( true );

			HibernateOrmScopeSessionContext sessionContext = mappingContext.sessionContext( upperSession );
			ictx.contextData().put( MassIndexingSessionContext.class, sessionContext );

			ictx.invoke( next -> {
				SessionImplementor session = sessionContext.session();
				ictx.contextData().put( SessionImplementor.class, session );
				try {
					beginTransaction( session, transactionTimeout );
					next.proceed();
					session.clear();
				}
				finally {
					// it's read-only, so no need to commit
					rollbackTransaction( session );
				}

			} );

		}
	}



```

Reply:
The interception mechanism when loading entities for indexes can be treated generally. The same problem arises when loading an entity. You can resolve it with the implementation "org.hibernate.search.mapper.pojo.loading.LoadingInterceptor". The same "TransactionInterceptors.withJTATransaction (tm)", can handle transactions while loading an entity. For further use in the new PR, when building the index and entity loading configurator for mapper-javabean. 

Exaple:
```java
options.registerLoader( Book.class, (identifiers) -> {
	return ....;
} ).arroundInvoke( TransactionInterceptors.withJTATransaction( tm ) );
```